enhancements/cluster-logging/loki-storage.md

```diff
@@ -0,0 +1,628 @@
+---
+title: Cluster Logging Loki as an alternative Log Store
+authors:
+  - "@periklis"
+reviewers:
+  - "@ewolinetz"
+  - "@blockloop"
+  - "@lukas-vlcek"
+  - "@sichvoge"
+approvers:
+  - "@ewolinetz"
+  - "@alanconway"
+  - "@jcantril"
+creation-date: 2021-03-29
+last-updated: 2021-03-29
+status: provisional
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Cluster Logging: Loki as an alternative Log Store
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The purpose of this is to provide an alternative low-retention log store based on [Loki](https://github.com/grafana/loki) for the cluster logging stack. This aims to provide a cloud-native first solution specialized in storing logs and in turn lessen the operational cost running cluster logging.
+
+Currently our main log store, i.e. Elasticsearch, is a more general-purpose ingestion, indexing and querying cluster-based solution for text search. It is a powerful solution with fine-grained control on how to scale log storage, indexing and querying.
+
+In general it serves its purpose quite well but it is a constant source of labor regarding operations. Its powerfulness and complexity requires a constant toll on the human operator side to understand how to individually scale the cluster as a log store. Representative actions that a human operator needs to figure out besides controlled operations by [elasticsearch operator](https://github.com/openshift/elasticsearch-operator):
+- Regular evaluation and balance of the CPU/Memory requests and limits because configuration for ingestion and query paths is not individual configurable.
+- Regular evaluation and adaption of redundancy effects on memory and persistent volume usage.
+- Usage of `managementState: Unmanaged` to unblock the cluster in situations of bugs, e.g. locked indices, shard allocation.
+- Usage of the Elasticsearch cluster API to tune the cluster for options not supported in the Elasticearch custom resource.
+
+On the other hand and based on our experience operating Loki as an internal central log aggregation solution, we found that a more specialized solution preserves the same log ingestion/querying user experience while lowering the human operational burden. Representative characteristics that showcase the improvements:
+- All components are stateless and can be scaled out individually based on load scenarios.
```

Comment:
Loki is a log *store*, so *all* the components can't be stateless.

Reply:
Loki is run stateless aka all it's components can be simple deployments w/o PVCs. The statefulness part is offloaded to the object storage, e.g. S3.

Reply:
One exception to the rule is the ingester only if you run it with the Write-Ahead-Log feature enabled.

Reply:
@syedriko Loki is not a "log store." By its own definition, it is a "log aggregation system." The storage is separate from Loki itself as Peri mentioned. The short-term statefulness of Loki can be enabled for efficiency, but it is not required by any means since the storage is offloaded to S3.

Reply:
@blockloop Loki is log store from the point of view of the user of the cluster logging stack, in the sense that you say 
 logStore:
    type: "loki" 
flip the power switch off, flip it back on and expect to find your logs where you last saw them.
My point being, Loki vs Elasticsearch it is not a fair comparison, the comparison needs to be between Loki + S3 backend vs Elasticsearch. On public cloud, administering the S3 backend can be made someone else's problem - for a price, but that's not all the cases out there.

Reply:
I'm not sure I understand the reason for the distinction in the language. Your original comment appeared to be aimed at misrepresenting the statelessness of the stack when, in fact, the stack itself is stateless - save for any outside dependencies. The comparison between Elasticsearch and Loki is more superficial than the storage device itself. Ad absurdum, the comparison can technically delve into PVC storage engine vs S3 which doesn't provide anything to the conversation this enhancement aims to have. At least I don't think so. I think the point of this paragraph is to compare the two objective realities of the stacks. That being said, the Loki stack that _we control_ via the operator is entirely stateless and can be deployed as such, regardless of the syntactical differences we admit in calling it a store or an aggregation engine.

Reply:
@blockloop The very title of this document calls Loki a log store. We better come to some agreement on what it is.
I'm referring to the paragraphs just above that describe all the trouble one has to go to to administer Elastic as well as describing Loki as almost trouble free. This is what I think is not a fair comparison. In both cases it is expected that logs will be stored in highly available storage. In case of Elastic, it provides that highly available store, which is where many of the complexities of its administration originate. Loki does not provide its own highly available store but rather relies on one behind an S3 interface. But that object storage needs provisioning and administration which is likely non-trivial. It can be outsourced, it can be amortized over many consumers, but to completely eliminate it from the picture I think is incorrect. Let's say we're deploying OCP on bare metal and want to install a Loki based cluster logging solution on top of that. Would we need to deploy an S3-compatible store (Ceph?) even if we don't need it for anything else?

Reply:
I think I got it - look at this not from the perspective of writing an operator to run Loki but from the perspective of providing log storage for the Openshift cluster logging stack. Remove Elastic from the current stack and replace it with Loki - will there be a gap left?

Reply:
Folks thank you first of all for this lively discussion about the details. IC where @syedriko drives the conversation. However, I would like to remind a couple of things to drive this to something useful like amending this section (please propose your amendments):
1. This is a summary section, it is not meant to be an exhaustive fair comparison between Loki and Elasticsearch. The effort here is to make visible the administrative burden on the one hand while proposing the mission to relief on the other hand.
2. The most important keyword is not `object storage` per se but `scale out`. For sure you can scale Elasticsearch horizontally. Its statefulness to do so is something one needs to master. The persistent volume attached is only one visible part of this statefulness. Other less visible to the non-expert are the shard allocation and the impact on memory for keeping in memory those shards to answer queries.
3. This section aims to highlight that you can scale the two paths ingestion and querying individually. This seems to be easier to achieve with dedicated individually scaled components like in Loki rather than role-based cluster nodes like in Elasticsearch. The separation of concerns is an indicator pro Loki here, but experience will teach us best.
4. This proposal is not encompassing all the details about object storage support by purpose. The message here is that the API supports to select the S3 storage of your choice. IMHO the support matrix of Loki and various object storage providers (S3, GCS, OpenShift Storage, etc.) belongs to a separate enhancement proposal that aims to close the loose ends from an Cluster-Logging API perspective. Our `ClusterLogging` API is unfortunately a bit biased using fields like `logStore` but this is something to be addressed there and not here. This proposal make the trade-off to re-use `store` as wording for Loki for the sake of simplicity.

Reply:
@periklis As far as this text goes, it may be best to just drop the entire Summary section. What it describes clearly does not fit every OCP target environment and is, frankly, light on detail and not convincing. As far as motivation, we're adding Loki for those use cases where it is appropriate. I would also suggest dropping "alternative". From the point of view of the cluster logging stack, it's a log store. There may be others in the future. What I would add is how log collectors are going to talk to Loki (I'm assuming we're keeping fluentd and adding a Loki plugin to it) and how the UI is going to be exposed (I'm assuming Kibana is coupled with Elastic).

Reply:
I think there is a slight disconnect with verbiage happening...

The purpose of this proposal is to talk about the API for the Loki Operator of which it provides a Loki cluster as a service to anything else that would use it. CLO is one of those customers however it is not the sole one.

To the Loki Operator it is a log aggregation system. There are multiple components that make up the Loki solution that it is responsible for managing and keeping healthy so that any of the clients that interact with Loki just have a simple end point to push entries to or query from. It allows us to give you the appearance it is a place to send your logs to or query them from, however it is not a simple "logstore".

It seems like this may be something we need to add to the description to clarify things.
Maybe something like "what Loki is/isn't" and we can resolve this thread?

Reply:
@ewolinetz I think we need to clarify the scope, if it is just about the API for the Loki operator or if is it about plugging in a log store based on that Loki operator into the larger logging stack.

Reply:
@syedriko I really appreciate that you widen the horizon on this topic with interesting questions, e.g. drop alternative wording (I assume you mean we keep both stores supported. That's also a question for @sichvoge), UI exposure. In general I agree these topics need some sort of resolution. However, the summary block in this doc has some important constraints.

First of all the section is mandatory for our enhancement proposal process. It cannot be simply dropped. The summary block is not meant to be a management summary for a whole project or epic. The purpose is clearly to provide a entrypoint/context to the text below according to the [template](https://github.com/openshift/enhancements/blob/master/guidelines/enhancement_template.md#summary). It should not be an exhaustive description of what a team wants to achieve across topics that require multiple enhancement proposal docs (e.g. UI expose aka replacing Kibana maybe?)

I have to admit that I have a hard time reading this thread to identify amendments that clearly shape the document in a better form. The summary provided aims to provide an entrypoint on why we want to provide a alternative store regardless if the alternative means replacement or not. The agenda to keep/drop Elasticsearch is a topic for another proposal or arch meeting. In addition the summary uses a light comparative approach to highlight differences, e.g. statefulness/statelessness, individual/combined scale out of ingestion and querying. It does neither compare in detail all the aspects nor tries to put Elasticsearch in a dark corner. The reader should be informed by this section on how to calibrate his lenses on a highly overloaded term as `logStore`.

Finally, the document is scoped to a `LokiStack` API that is provided by the `loki-operator`. For the sake of a useful bridge to the `ClusterLogging` API, this proposal also describes how to achieve an integration with present means and no more.

Reply:
@syedriko would this be clear/complete?:
- The scope here is the API to Loki operator
-  Logging stack users that want Loki as a drop-in replacement to ES would just set `logStore: loki` in ClusterLogging, the operators take care of the rest (do we need to expose some S3 info in ClusterLogging for this to work??)
-  Logging stack users that want more control/a different Loki instance see https://github.com/openshift/enhancements/pull/710#pullrequestreview-631355181

We can add a note to each of the proposals to that effect.

Reply:
I guess @syedriko has a point that somewhere we have to describe the changes to ClusterLogging to enable Loki. I'm inclined to add that to https://github.com/openshift/enhancements/pull/710#pullrequestreview-631355181 rather than here
- it's not worth creating another proposal just for that
- its healthy for this to describe a Loki operator independently of how CLO will use it.
- #710 is already about configuring the CLO,  makes sense to discuss CL alongisde CLF